### PR TITLE
Circuit cloning balance: Timed cloning, reduced upgrade disk material cost

### PR DESCRIPTION
**Pull request by ghost**
:cl: Xhuis
balance: Instead of starting unable to clone circuits at all, circuit printers can now print circuits over time from roundstart. The formula for this is equal to (metal cost / 150) seconds, with a maximum of 3 minutes. You can see printing progress by using the printer's interface, and you can print normal components during this time.
balance: If circuit printing is disabled in the config, cloning remains unavailable.
balance: The upgrade disk to allow circuit printers to clone circuits has been replaced with an upgrade disk to make circuit cloning instant.
balance: Both circuit printer upgrade disks now cost 5000 metal and glass, down from 10000.
/:cl:

Having messed with circuits a lot lately, on both servers, I can say that Sybil was by far the less enjoyable one. Without cloning, making use of the toys you spend a ton of time working together is essentially impossible because of how fast rounds are, but it also has the balance problem of allowing you to mass-produce highly powerful circuits as soon as the round starts. (I'm not arguing that cloning should be enabled on Sybil, bear with me here.)

This tries to strike a balance by implementing printing over time. Instead of starting with or without the ability to clone circuits, printers (by default) now have *slow* cloning. Entering a circuit code and printing will consume the materials immediately and slowly begin creating the assembly over a number of seconds equal to `metal cost / 150`, with a maximum time of three minutes to clone. Once it finishes, the assembly will be created. You can cancel printing at any time for a metal refund.

Advanced parts are still locked behind the upgrade disk.

Hopefully, this will allow people to make use of their goodies without forcing them to either tell R&D to get advanced data soon or making them hate you by immediately rushing it at roundstart.

I also reduced the costs of the two upgrade disks by half; instead of 10k metal and glass, they now cost 5k. Additionally, I increased the circuit printer metal cost to 7.5k for sensibility's sake. This is for two reasons: one, because they costed twice as much as an actual printer, which is silly; and two, because they served as a resource sink that isn't necessary because if you're running low on the resources you either have to cannibalize the station or ~~steal~~ borrow them from somewhere else, which is also silly. (I'm open to remove this if support for the main PR is high but not the cost reduction.)

### CODE REVIEW

**Comment by kevinz000** [comment]
remember when we introduced laser miniguns to science to take out the smg and then removed it because it was op?
well welcome to circuits
i'm not surprised though i knew this would happen it's literally only a matter of time

**Comment by Xhuis** [comment]
what did he mean by this

**Comment by MMMiracles** [comment]
the circuit printer is like a laser minigun but instead of spewing out billions of lazers it spews out billions of robots that scream "ASS".

**Comment by Xhuis** [comment]
Hi, I was that RD. I brought two laser guns with me, and there were two non-strong blob tiles between the core and space. I did bring some recharger circuits with me to fuel my guns, but had it rallied spores or defended its core better then I find it likely that it could have survived, given that I was alone.

The fact that you're already talking about nerfing them is incredibly funny, because in deadchat that round I was commenting about how you were angry that your blob defense was interrupted and that you were probably going to nerf circuits because of it.

**Comment by arsserpentarium** [comment]
Mkay.

**Comment by Tlaltecuhtli** [comment]
wait for meme pa cells on all guns

**Comment by arsserpentarium** [comment]
it is instant now.But only with disk.

**Comment by ExcessiveUseOfCobblestone** [comment]
Ok so compared to that...

- now you get slowning roundstart
- now it takes less to get the instant disk

What is the compromise here? Those still look like straight buffs.

**Comment by Robustin** [comment]
Hey I explained I was heading away to refill my oxygen. Even lings need air. 

Like I said I don't even plan on looking at it yet but blob core have autogenerating strong blobs and 300hp, even an emitter firing directly on the core won't kill it if the blob is AFK. So my concern is with circuits basically creating infinite ammo lasers. 

The key questions for me are:

1) How long does it take for the recharger to recharge a depleted laser?

2) Does the laser recharge without having to remove it from your inventory?

3) Does the circuit ever run out of power? How many lasers can it recharge before it does? 

But that's a totally different issue than this PR so meh.

**Comment by Kromgar** [comment]
Lings don't need air if you use the flesh space suit

**Comment by Xhuis** [comment]
@Cobby It could be considered a buff, yes. Without at least some form of cloning from the get go, coupled with no advanced components, circuits aren't really worth using until the upgrades are acquired. I'm giving a thing to make them at least capable of making little gadgets without building them from scratch each time.

**Comment by arsserpentarium** [comment]
answers for Robustin:
1.It depends on how good will be your assembly and what laser do you have.in best case about 20 s to ull charge.
2.It can't recharge laser in your backpack if it doesn't in backpack too.
3.Yes.If there is no tesla relay, it will just transfer all internal battery energy to laser with 80% efficiency.
also circuits can recharge from some reagents and powernet.But these methods have certain limitations.

**Comment by ExcessiveUseOfCobblestone** [comment]
@Xhuis I haven't seen them in action enough to say with good faith if the PR is good/bad

I'm just not sure how straight buffs is "compromising" with I'm guessing complete removal of instant cloning. Imagine trying to average 0 and 100 and you end up with 125. 

Can you perhaps re-title your PR to something a bit less misleading?

**Comment by Xhuis** [comment]
The compromise is to give printers a basic cloning function without removing it entirely. It was recently completely removed without upgrading it first. I think the PR title is fine.

**Comment by epochayur** [comment]
how is the title misleading?

**Comment by ExcessiveUseOfCobblestone** [comment]
A compromise is "half way" (give or take in the world of coderbus). This isn't remotely "half way", it's exactly how it was before PLUS slow-ning and a lower material cost.

"people want cloning removed entirely so ill _compromise_ by adding slow cloning and making the instant cloning cost less!"

I don't think it's unfair of me to ask you retitle it so that it's clear nothing is being done in this PR to actually appease people who want it removed entirely, and in fact this PR does quite the opposite by adding cloning roundstart (albeit slower).

**Comment by arsserpentarium** [comment]
mkay, let's make upgrade disk not make cloning instant, but faster?

**Comment by arsserpentarium** [comment]
I doubt, that FAST assembly production is really needed.
But we really need some compromise,to avoid attacks of technophobes on vital features for our secret society of circuitrynerds.To let us autize on new constructions.

**Comment by ExcessiveUseOfCobblestone** [comment]
If the buff is good for the game then you could just retitle the pr lol,

**Comment by Xhuis** [comment]
If you're this triggered over a single word, I'd hate to see you in a political debate.

**Comment by ExcessiveUseOfCobblestone** [comment]
If you are trying to call your way +5 a "compromise", I'd hate to see you in politics at all.

**Comment by Xhuis** [comment]
This PR serves to do a middle ground between "no cloning" and "instant cloning." A middle ground is a compromise.

File: code/modules/integrated_electronics/core/printer.dm [hunk 1]
```diff
@@ -16,11 +21,28 @@
...
+	if(!cloning)
+		STOP_PROCESSING(SSprocessing, src)
+	clone_countdown--
+	if(!clone_countdown || fast_clone)
+		var/turf/T = get_turf(src)
+		T.visible_message("<span class='notice'>[src] has finished printing its assembly!</span>")
+		playsound(get_turf(T), 'sound/items/poster_being_created.ogg', 50, TRUE)
+		SScircuit.load_electronic_assembly(T, program)
+		cloning = FALSE
+		STOP_PROCESSING(SSprocessing, src)
```
Comment (Cyberboss) [inline code/modules/integrated_electronics/core/printer.dm]: return PROCESS_KILL

File: code/modules/integrated_electronics/core/printer.dm [hunk 2]
```diff
@@ -215,14 +242,38 @@
...
+					if(!materials.use_amount_type(program["metal_cost"], MAT_METAL))
+						to_chat(usr, "<span class='warning'>You need [program["metal_cost"]] metal to build that!</span>")
+						return
+					var/cloning_time = program["metal_cost"] / 150
+					cloning_time = min(cloning_time, MAX_CIRCUIT_CLONE_TIME)
+					cloning = TRUE
+					clone_countdown = cloning_time
+					to_chat(usr, "<span class='notice'>You begin printing a custom assembly. This will take approximately [round(cloning_time / 60, 0.1)] minute(s). You can still print \
+					off normal parts during this time.</span>")
+					playsound(get_turf(src), 'sound/items/poster_being_created.ogg', 50, TRUE)
```
Comment (Cyberboss) [inline code/modules/integrated_electronics/core/printer.dm]: playsound(src, ...

File: code/modules/research/designs/electronics_designs.dm [hunk 3]
```diff
@@ -57,31 +57,31 @@
...
 
 /datum/design/IC_printer_upgrade_clone
-	name = "Integrated Circuits printer upgrade: Clone Ability"
-	desc = "This disk allows for integrated circuit printers to clone designs."
+	name = "Integrated circuit printer upgrade: Instant Cloning"
+	desc = "This disk allows for integrated circuit printers to clone designs instantaneously."
 	id = "icupgclo"
 	build_type = PROTOLATHE
-	materials = list(MAT_GLASS = 10000, MAT_METAL = 10000)
+	materials = list(MAT_GLASS = 5000, MAT_METAL = 5000)
```
Comment (Cyberboss) [inline code/modules/research/designs/electronics_designs.dm]: Halving the cost seems like pushing it too far

File: code/modules/integrated_electronics/core/printer.dm [hunk 4]
```diff
@@ -215,14 +242,38 @@
...
+				else
+					var/datum/component/material_container/materials = GetComponent(/datum/component/material_container)
+					if(!materials.use_amount_type(program["metal_cost"], MAT_METAL))
+						to_chat(usr, "<span class='warning'>You need [program["metal_cost"]] metal to build that!</span>")
+						return
+					var/cloning_time = program["metal_cost"] / 150
+					cloning_time = min(cloning_time, MAX_CIRCUIT_CLONE_TIME)
+					cloning = TRUE
+					clone_countdown = cloning_time
+					to_chat(usr, "<span class='notice'>You begin printing a custom assembly. This will take approximately [round(cloning_time / 60, 0.1)] minute(s). You can still print \
```
Comment (Cyberboss) [inline code/modules/integrated_electronics/core/printer.dm]: no backslashes mid string

File: code/modules/integrated_electronics/core/printer.dm [hunk 5]
```diff
@@ -33,11 +55,11 @@
 		return TRUE
 
 	if(istype(O, /obj/item/disk/integrated_circuit/upgrade/clone))
-		if(can_clone)
+		if(fast_clone)
 			to_chat(user, "<span class='warning'>\The [src] already has this upgrade. </span>")
 			return TRUE
-		to_chat(user, "<span class='notice'>You install \the [O] into \the [src]. </span>")
-		can_clone = TRUE
+		to_chat(user, "<span class='notice'>You install \the [O] into \the [src]. Circuit cloning will now be instant. </span>")
```
Comment (Cyberboss) [inline code/modules/integrated_electronics/core/printer.dm]: `\the`s unecessary

File: code/modules/integrated_electronics/core/printer.dm [hunk 6]
```diff
@@ -16,11 +21,28 @@
...
 
+/obj/item/device/integrated_circuit_printer/Destroy()
+	STOP_PROCESSING(SSprocessing, src)
+	return ..()
+
+/obj/item/device/integrated_circuit_printer/process()
+	if(!cloning)
+		STOP_PROCESSING(SSprocessing, src)
+	clone_countdown--
+	if(!clone_countdown || fast_clone)
```
Comment (Cyberboss) [inline code/modules/integrated_electronics/core/printer.dm]: this can be achieved in a much more performant way using stoppable timers

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/modules/integrated_electronics/core/printer.dm (hunk 1 above)
Comment (Xhuis) [reply code/modules/integrated_electronics/core/printer.dm]: 🔪

Comment (Cyberboss) [reply code/modules/integrated_electronics/core/printer.dm]: See my last comment

Comment (Xhuis) [reply code/modules/integrated_electronics/core/printer.dm]: @Cyberboss 🔫 😏

File: code/modules/research/designs/electronics_designs.dm (hunk 3 above)
Comment (Iamgoofball) [reply code/modules/research/designs/electronics_designs.dm]: Metal and glass are infinite. They don't matter.

File: code/modules/integrated_electronics/core/printer.dm (hunk 6 above)
Comment (Xhuis) [reply code/modules/integrated_electronics/core/printer.dm]: Would that be by using something like
`clone_countdown = addtimer(blah etc. blah)`
and then deleting that timer if canceled?

File: code/modules/integrated_electronics/core/printer.dm (hunk 2 above)
Comment (Xhuis) [reply code/modules/integrated_electronics/core/printer.dm]: While I'm on the subject, when should I use one or the other? I've always generally just used `src` but nowadays I'm called out seemingly at random.

**Comment by KorPhaeron** [comment]
I dont know enough about circuits to review this

**Comment by optimumtact** [comment]
fix

**Pull request merged**